CASSANDRA-20830 AST Fuzz testing support for tables w/ mutation tracking enabled and single-partition reads

maedhroz opened the pull request:
patch by Caleb Rackliffe; reviewed by ? for CASSANDRA-20930

maedhroz committed:
Enable single-partition AST fuzz tests w/ mutation tracking

patch by Caleb Rackliffe; reviewed by ? for CASSANDRA-20830

dcapwell commented on test/distributed/org/apache/cassandra/distributed/test/cql3/MultiNodeTableWalkWithMutationTrackingTest.java:
```diff
@@ -0,0 +1,53 @@
+        // CQL_DEBUG_APPLY_OPERATOR = true;
+        // When mutations look to be lost as seen by more complex SELECTs, it can be useful to just SELECT the partition/row right after to write to see if it was safe at the time.
+        // READ_AFTER_WRITE = true;
+    }
```
sounds like we should add

```
IGNORED_ISSUES.remove(KnownIssue.AF_MULTI_NODE_MULTI_COLUMN_AND_NODE_LOCAL_WRITES);
```

That would test if [CASSANDRA-19007](https://issues.apache.org/jira/browse/CASSANDRA-19007) is in fact fixed by MT

dcapwell reviewed (approved):
LGTM, i know its not passing but the test looks fine

bdeggleston committed:
Single-partition read fixes:
- filter augmenting mutations and add comment explaining why we don't exclude reconciled mutations when augmenting
- propagate exceptions during acknowledgeReconcile
- fix UnreconciledMutations handling of single key bounds
- fix single offset remove bug

patch by Blake Eggleston; reviewed by Caleb Rackliffe for CASSANDRA-20830

aratno committed:
Fix loss of isRepaired StatsMetadata on compaction race

Found via MultiNodeTableWalkWithMutationTrackingTest after ~hundreds of
seeds.

java.lang.IllegalArgumentException: Cannot compact repaired and unrepaired sstables
    at com.google.common.base.Preconditions.checkArgument(Preconditions.java:143)
    at org.apache.cassandra.db.compaction.AbstractCompactionTask.validateSSTables(AbstractCompactionTask.java:84)
    at org.apache.cassandra.db.compaction.AbstractCompactionTask.<init>(AbstractCompactionTask.java:60)
    at org.apache.cassandra.db.compaction.CompactionTask.<init>(CompactionTask.java:79)
    at org.apache.cassandra.db.compaction.CompactionTask.<init>(CompactionTask.java:74)
    at org.apache.cassandra.db.compaction.LeveledCompactionTask.<init>(LeveledCompactionTask.java:39)
    at org.apache.cassandra.db.compaction.LeveledCompactionStrategy.getMaximalTasks(LeveledCompactionStrategy.java:192)
    at org.apache.cassandra.db.compaction.CompactionStrategyHolder.getMaximalTasks(CompactionStrategyHolder.java:121)

patch by Abe Ratnofsky; reviewed by Caleb Rackliffe for CASSANDRA-20830
